### ISSUE TITLE
Please can I have prebuilt binaries for simh?

### ISSUE THREAD

**Issue by MidnightThunder** [open-simh/simh#54]
I can't get it to compile by myself and compiling code is not something i really understand, could I please have prebuilt binaries for windows?

**Comment by pkoning2** [comment]
Yes, that's a goal.  It needs someone conversant with Windows to do the necessary work.

**Comment by Xuser641** [comment]
So at the moment it does not work with Windows, correct?

**Comment by pkoning2** [comment]
I haven't tried it, but I assume it does work.  The missing part as far as I know is only the prebuilt binaries, not the ability to run.

**Comment by LegalizeAdulthood** [comment]
I believe we should be able to include pre-built binaries from the continuous integration workflow into the releases here on github.

**Comment by pkoning2** [comment]
Interesting idea.  Is that enough for what is needed?  For the Windows case, installers are typical, though I suppose a simple ZIP archive may be sufficient if nothing installer-flavored is needed.  That would be worth investigating.  Any volunteers?

**Comment by markpizz** [comment]
Well, every commit (or at least every push) produces the CI activity.  Each of those then produces almost 120MB Windows of binaries.  These can be compressed down to about 11MB for all the binaries, if that is stored for each relatively minor commit it seems like it's quite wasteful.

I think it would be better to drive the release concept packaging on some sort of calendar process (at most quarterly maybe).

**Comment by pkoning2** [comment]
Good point.  But the other part of the suggestion was that the CI process is all that is needed to produce the ingredients for kits.  Is that true?

**Comment by LegalizeAdulthood** [comment]
Starting with a ZIP file is an easy place to start; an installer can come later.  It's not hard to do with CMake but would be a giant PITA with the existing makefile, unless it was already there.  (I haven't looked.)

**Comment by LegalizeAdulthood** [comment]
It wouldn't necessarily be producing a package per CI build, it would be leveraging the CI build to create a package associated with a release, either manually added to the release or somehow automated into the release based on a manual trigger.  My point was that we have automated Windows stuff building via the CI infrastructure, so you're about 80-90% of the way towards an automated release mechanism for Windows.

**Comment by markpizz** [comment]
Maybe if the CI build process for Windows produces XP capable binaries (not yet with cmake), then maybe some fancy CI scripting could even automatically determine if the reasonable "release" interval had been passed and then push a formal release.

**Comment by LegalizeAdulthood** [comment]
IMO, making Windows XP a requirement is dumb.  If people want XP binaries they can build them from source.  You should target the common case for Windows users, which hasn't been XP for a long, long time (many years).

**Comment by pkoning2** [comment]
Well, if the tools can do it (which is supposed to be the case) then why not?  We're all about obsolete systems, after all.  And my WinXP system works just fine (for Microsoft definitions of "works" and "fine" that is).  I even have a Win95 kit somewhere I can dust off if I really want to go old style, but I wouldn't try to run SIMH on one of those.

**Comment by XVilka** [comment]
We use packaging based on tags and branch names. E.g., on `0.4.0` or `release-*` or `dist-*` (branch names helps to send pull requests generating these artifacts for all platforms). You could take a look at our CI scripts as an inspiration:
- https://github.com/rizinorg/rizin/blob/dev/.github/workflows/ci.yml
- https://github.com/rizinorg/cutter/releases/tag/v2.1.2

This generates releases like (except release notes, of course):
- https://github.com/rizinorg/rizin/releases/tag/v0.4.1
- https://github.com/rizinorg/cutter/releases/tag/v2.1.2

For Windows, we also have AppVeyor to do slightly more different configurations:
- https://github.com/rizinorg/rizin/blob/dev/.appveyor.yml

As for Windows XP, you could just use VS2015 as the latest one able to generate WinXP-compatible binaries.

**Comment by bscottm** [comment]
@XVilka, @pkoning2, @LegalizeAdulthood, @Xuser641: Go over to my fork of the open-simh repo, bscottm/open-simh. There are two releases: simh-4.0.0-alpha1 and simh-4.0.0-alpha2. There are several installation artifacts there: .exe and .msi for Windows 10 native, a ZIP for Windows XP and .deb for the two flavors of Ubuntu (22.04 and 20.04). I'm still cogitating what needs to be in a Mac OSX `plist` file, which is why there is no Mac OSX package yet.

The Github CI/CD process will generate a release if the repo is tagged with a tag starting with "v" or "simh-". It's all handled by the GH Actions, everything is built by CMake-generated build systems (Ubuntu -> ninja, Mac OSX -> xcodebuild, Windows 10 native -> msbuild, WinXP -> msbuild), the installers are created by CPack. The CMake-build process is a reusable workflow.

End-to-end CMake.

The Windows installers are unsigned, so they have to be installed manually from a PowerShell or cmd console:

```
# Install using the NSIS-generated .exe:
> .\simh-4.0.0-win32-native.exe

# Install using the WIX-generated .msi:
> msiexec /qf /i simh-4.0.0-win32-native.msi
```

There are known defects:
- `microvax3900` is not installed or linked to `vax`. That requires installer customization that isn't relevant to getting a minimum viable product off the ground.
- The executables are not prefixed with `simh-`. That would introduce a breaking change for the existing developer community that needs to be discussed. CMake does not rename executables at install time.

I fixed the PDP-8 issue (it was not in the DEC PDP simulator family) and the Python3 generator script collects the list of simulators in a particular "family", adds them to the family's description.

@pkoning2: The Python3 generator script sanity checks the list of simulators to ensure none are missing. I grabbed the list from the `all` and `exp` makefile rules. The Python3 script will "Mark Pizzolato" if any are missing. None are missing.

**Comment by XVilka** [comment]
Now, there with every successfully built commit: https://github.com/open-simh/simh/actions/runs/5009122925

**Comment by bscottm** [comment]
Is that a good or a bad thing?

**Comment by XVilka** [comment]
> Is that a good or a bad thing?

Obviously a good one. It is also possible to build "nightly" releases, like NeoVim does, as an alternative: https://github.com/neovim/neovim/releases
